### ISSUE TITLE
Idea: Add support for Kotlin Properties.

### ISSUE THREAD

**Issue by omagarwal25** [amarcolini/joos#14]
Currently, when I have to do a config class, I have to do something like this

```kt
 @Config
    object Coefficients {
        @JvmField
        var MODULE_PID = PIDCoefficients(50.0, 1.0, 2.0)

        @JvmField
        var TRAJECTORY_CONSTRAINTS = DiffSwerveConstraints()

        @JvmField
        var TRANSLATIONAL_PID = PIDCoefficients(1.0, 0.0, 0.5)

        @JvmField
        var HEADING_PID = PIDCoefficients(1.0, 0.0, 0.5)
    }
```

It would be nice if I could forgo the @JvmField annotation and it would still add them to the dashboard. If this behavior is possible that would be great, but if not no worries.

Also question. What does WithConfig change compared to the already existing ability to modify class instances in the dashboard? I've tuned PID using this, but I'm curious what the new release and annotations change.

**Comment by amarcolini** [comment]
So it should be possible to forgo the @JvmField annotation by using [ksp](https://kotlinlang.org/docs/ksp-overview.html) instead of kapt, but I haven't implemented that yet. 

`WithConfig` allows for custom config behaviors for a specific class. For example,
```kotlin
@WithConfig(House::class)
class House(var rooms: Int, var doors: Int) : ConfigProvider<House> {
    override fun parse(house: House) = CustomVariable().apply {
        putVariable("rooms", BasicVariable(object : ValueProvider<Int> {
            override fun get(): Int = rooms

            override fun set(value: Int?) {
                if (value == null) return
                house.rooms = value
                house.doors = value + 1
            }

        }))
    }
}
```
Here the `House` class would not normally have any variables in its config (because Joos doesn't currently scan kotlin types), but with the WithConfig annotation, `House` is used as a config provider, so whenever a `House` is used as a config variable, the custom config will be used. Also, the class specified in `WithConfig` doesn't have to be the same class, I just did that in this example.

Basically the `WithConfig` annotation allows for custom behaviors, i.e., having the number of doors be one greater than the number of rooms, or in the case of `Angle`, using the default units when displaying the config, etc.

I could also add more features like an annotation to exclude fields from being config variables, or being able to add the `Config` annotation to individual fields, if you think it's a good idea.

**Comment by omagarwal25** [comment]
Ah. I understand that makes sense. Oh, I also tested the IMU code. It seems to be working correctly for heading even though the auto-detect up axis function is returning the wrong axis.

**Comment by amarcolini** [comment]
What's the orientation of your IMU and the orientation that autoDetectUpAxis returned?

**Comment by omagarwal25** [comment]
It's vertical. It returns Y.

**Comment by amarcolini** [comment]
*
     *                           | Z axis
     *                           |
     *     (Motor Port Side)     |   / X axis
     *                       ____|__/____
     *          Y axis     / *   | /    /|   (IO Side)
     *          _________ /______|/    //      I2C
     *                   /___________ //     Digital
     *                  |____________|/      Analog
     *
     *                 (Servo Port Side)
  Using this diagram, which axis is pointing up (or down)?

**Comment by omagarwal25** [comment]
X up. Servo ports are on the bottom.

**Comment by amarcolini** [comment]
I'll try and fix it in the next release. By the way, how did your competition go?

**Comment by omagarwal25** [comment]
Two weeks to go. It's all just code now. Joos has really helped with auto.

**Comment by amarcolini** [comment]
Glad I could help!

**Comment by amarcolini** [comment]
Version 0.4.5-alpha should address (and hopefully fix) these issues, if you'd like to try it out.

**Comment by omagarwal25** [comment]
Oh also, `@JoosConfig` does not work. I had to switch back to regular `@Config`. at least in 0.4.4

**Comment by amarcolini** [comment]
Does no config show up at all? How did you install it?

**Comment by omagarwal25** [comment]
I used kapt. I have switched to KSP. I'll let you know how it goes in maybe two hours.

**Comment by amarcolini** [comment]
I see why it wouldn't work. I'll fix it as soon as I can.

**Comment by omagarwal25** [comment]
IMU returns the right axis now.

**Comment by amarcolini** [comment]
Good to hear! The annotation processor should be fixed in 0.4.5-alpha-patch now.

**Comment by omagarwal25** [comment]
I'll test it soon.

**Comment by omagarwal25** [comment]
Still doesn't show up.

**Comment by omagarwal25** [comment]
Hey @amarcolini, thanks for all your help. We won the Innovate Award! We were also in the running for the control award.

**Comment by amarcolini** [comment]
Congratulations!! I'm so happy that you found Joos helpful. And by the way, I'm hoping to get access to some hardware soon so that I can test the annotation processor (and everything else) better.

**Comment by omagarwal25** [comment]
I was thinking over the summer I was going to write some sensor components to make life easier.

**Comment by amarcolini** [comment]
This is definitely fixed as of 0.4.9 (and possibly earlier). Sorry I forgot about this issue. Also, `WithConfig` has been replaced in favor of the `ImmutableConfigProvider` and `MutableConfigProvider` annotations.
